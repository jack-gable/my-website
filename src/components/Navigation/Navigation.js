"use client";
import React from "react";
import { motion } from "framer-motion";
import Link from "next/link";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
	faEnvelope,
	faFilePdf,
	faFileWord,
	faHome,
	faLaptopFile,
	faUser,
} from "@fortawesome/free-solid-svg-icons";
import styled from "styled-components";

const LINKS = [
	{
		slug: "home",
		label: "HOME",
		href: "/",
		icon: faHome,
		target: "",
	},
	{
		slug: "about",
		label: "ABOUT",
		href: "#about",
		icon: faUser,
		target: "",
	},
	{
		slug: "projects",
		label: "PROJECTS",
		href: "#projects",
		icon: faLaptopFile,
		target: "",
	},
	{
		slug: "contact",
		label: "CONTACT",
		href: "/contact",
		icon: faEnvelope,
		target: "",
	},
	{
		slug: "resume",
		label: "RESUME",
		href: "/",
		icon: faFilePdf,
		target: "_blank",
	},
];

export default function Navigation() {
	const [hoveredNavItem, setHoveredNavItem] = React.useState(null);

	const id = React.useId();

	return (
		<nav onMouseLeave={() => setHoveredNavItem(null)}>
			<NavList>
				{LINKS.map(({ slug, label, href, icon, target }) => (
					<ListItem
						key={slug}
						style={{
							zIndex: hoveredNavItem === slug ? 1 : 2,
						}}
					>
						{hoveredNavItem === slug && (
							<Hovered
								layoutId={id}
								initial={{
									borderRadius: 8,
								}}
							/>
						)}

						<NavLink
							href={href}
							target={target}
							onMouseEnter={() => setHoveredNavItem(slug)}
						>
							<FontAwesomeIcon icon={icon} size="lg" />
							<div>{label}</div>
						</NavLink>
					</ListItem>
				))}
			</NavList>
		</nav>
	);
}

const NavList = styled.ul`
	display: flex;
	gap: 16px;
	padding: 0;
	list-style-type: none;
`;

const ListItem = styled.li`
	position: relative;
`;

const NavLink = styled(Link)`
	display: flex;
	align-items: baseline;
	gap: 8px;
	position: relative;
	padding: 8px;
	text-decoration: none;
	color: var(--color-offWhite);
	opacity: 0.7;
	transition: opacity 500ms;

	&:hover,
	&:focus {
		opacity: 1;
		filter: brightness(110%);
		color: var(--color-white);
	}
`;

const Hovered = styled(motion.div)`
	position: absolute;
	inset: 0;
	background: var(--color-gray-600);
`;
