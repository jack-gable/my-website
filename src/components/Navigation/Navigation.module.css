.navList {
  display: flex;
  gap: 12px;
  padding: 0;
  list-style-type: none;
  flex-direction: column;
}

.listItem {
  position: relative;
}

.navLink {
  border: none;
  background: transparent;
  cursor: pointer;
  border-radius: 12px;
  display: flex;
  align-items: center;
  gap: 8px;
  position: relative;
  padding: 6px 10px;
  text-decoration: none;
  color: var(--color-text);
  transition: background 200ms;
}

.navLink:hover,
.navLink:focus {
  background: var(--color-decorative-100);
}

@media (min-width: 35rem) {
  .navList {
    flex-direction: row;
  }
}
