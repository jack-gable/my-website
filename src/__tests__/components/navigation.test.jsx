import React from "react";
import Navigation from "@/components/Navigation";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";

describe("<Navigation />", () => {
	it("renders the <Navigation /> component with text", () => {
		const { container, getByText } = render(<Navigation />);

		expect(getByText("ABOUT")).toBeInTheDocument();
		expect(getByText("PROJECTS")).toBeInTheDocument();
		expect(getByText("CONTACT")).toBeInTheDocument();
		expect(container.firstChild).toMatchSnapshot();
	});
});
