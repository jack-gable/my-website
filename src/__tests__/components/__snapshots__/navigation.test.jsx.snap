// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> renders the <Navigation /> component with text 1`] = `
<nav>
  <ul
    class="navList"
  >
    <li
      class="listItem"
    >
      <a
        aria_label="about"
        class="navLink"
        href="/about"
      >
        ABOUT
      </a>
    </li>
    <li
      class="listItem"
    >
      <a
        aria_label="projects"
        class="navLink"
        href="/projects"
      >
        PROJECTS
      </a>
    </li>
    <li
      class="listItem"
    >
      <a
        aria_label="contact"
        class="navLink"
        href="/contact"
      >
        CONTACT
      </a>
    </li>
  </ul>
</nav>
`;
